### ISSUE TITLE
Add SpongePowered-like mixins to GDScript

### ISSUE THREAD

**Issue by CammiePone** [godotengine/godot-proposals#1346]
**Describe the project you are working on:**
I'm working on a modloader for a voxel game of mine, and my goal with said modloader is to give modders as much power over my own code as I possibly can, while maintaining compatibility between mods.

**Describe the problem or limitation you are having in your project:**
While general hooks are good, they are limited to what I think to add, and the fact that I can't really add a hook for every single line of code. This could possibly even be extended to allow you to mixin into parts of the engine itself through GDScript, which could allow greater power to game developers, but I'm also not entirely sure how you would convert GDScript to C++ in order to do it.

I come from making mods for Minecraft, and after getting used to using mixins, it's hard to imagine not being able to use them. Unfortunately SpongePowered's mixins are for Java only.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
As the title says, I think something like SpongePowered's mixins in GDScript would be extremely powerful. They're very different from the mixins mentioned in #758. Basically what they do is they allow someone to make a mixin class or interface to safely inject code into specific areas, even if other mods inject into those same exact places. It uses annotations, the main five being `@Mixin`, `@Accessor`, `@Inject`, `@At`, and `@Shadow`.

The Mixin annotation specifies that a class or interface is a mixin and to what class you're wanting to mess with, the Accessor annotation allows you to create getters or setters for variables you normally don't have access to, the Inject annotation specifies a method to inject into as well as where using the At annotation (HEAD (top of the method), TAIL (bottom of the method), and more specific things like before or after a variable or method is called in a method and which one to place the injection at if there's multiple), and the Shadow annotation allows you to use a variable as if your mixin class was inside the specified class without needing to create it yourself. You're also able to create your own variables, local and not, as well as methods through mixins if you wish, and can also overwrite a method entirely (however obviously this would be a last ditch thing, as it could break anything trying to inject into the method you overwrote).

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
So let's say I had a method for handling the movement of my player, and a modder wanted to inject into there. They could do something like this:

```
@shadow
var velocity

@accessor(script = "res://scripts/entities/Player.gd", variable = "health")
func get_health()

@inject(script = "res://scripts/entities/Player.gd", method = "move", @At(value = "PRE_VAR", target = "velocity", ordinal = 1))
func multiply_move_vector():
    if get_health() < 5:
        velocity *= 0.2
```

This is a very simple example of someone specifying the path to the script they want to mess with, and gaining access to my `velocity` variable within the mixin itself, and using an accessor to create an empty method (these would usually be used within an interface due to limitations with Java, but would be doable in an implementation in GDScript without it since, you know, you're able to modify GDScript itself) that returns the value of the `health` variable, and can be used outside of the mixin if the class were to be extended or autoloaded. From there, I'm injecting just before the second time the variable `velocity` is called in the `move()` method and adding a check for if the player's health is less than 5, and if it is, to slow them down.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I actually don't believe so. If it could, I would love to be told how to for when 4.0 hits. After I saw annotations were coming to GDScript, I immediately thought about how this could solve my conundrum for how to make a useful modding API without needing to add hooks myself into everything and then maintain those hooks.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Because I don't think it's possible without engine-level access. It could probably be done with a module, but I feel like if implemented in a way that allows users to inject code into the base engine as well, it could be extremely useful to people that don't know C++, but want to make a change to the engine's code that wouldn't otherwise be changeable, and it would definitely help with making mods for games that use Godot if a developer wants to build a modding community around their game.

**Comment by Calinou** [comment]
Related to (duplicate of?) https://github.com/godotengine/godot-proposals/issues/758.

**Comment by CammiePone** [comment]
Already mentioned it, and confirmed with a friend who knows D-like mixins and SpongePowered mixins well. The two aren't remotely similar, according to him. SpongePowered mixins handle code injections at runtime, kinda like ObjectWeb ASM (it's actually what it uses), but cleaner and easier to use, and with less chances for someone writing a mixin to mess it up.

**Comment by willnationsdev** [comment]
There was already a fully fleshed out plan for how to add a Trait system (i.e. a Mixin system) to GDScript outlined in godotengine/godot#23101. It just hasn't been ported to a formal Proposal yet. But vnen, the guy writing GDScript 2.0, was a very vocal/active user in that discussion and likely already has intentions of eventually adding the features we discussed there to GDScript - perhaps after he finishes the GDScript-to-C compilation tools or something.

That said, I'm not sure I like the mixin system you've proposed above.

> This could possibly even be extended to allow you to mixin into parts of the engine itself through GDScript, which could allow greater power to game developers, but I'm also not entirely sure how you would convert GDScript to C++ in order to do it.

This would require extensive changes to the scripting system in core as well as an arbitrarily large number of new hooks littered throughout the core code...without those things actually giving any utility without a module installed. I don't really see other core devs liking that kind of bloat being added to the core.

---

The rest of the feedback is comparing your Proposal to the way our old Issue proposed to do things.

> The Mixin annotation specifies that a class or interface is a mixin and to what class you're wanting to mess with

Sensible. We would likely have had a `trait` keyword at the top of a script, or perhaps a `@trait` annotation, similar to what has happened to `tool`/`@tool` in 4.0.

> the Accessor annotation allows you to create getters or setters for variables you normally don't have access to

All scripted properties are already public by default, so there isn't any reason to do this. And I don't think it's even *possible* to arbitrarily grant access to private member variables of C++ code without doing some crazy magical runtime compilation and hot reloading on the target machine which....seriously wouldn't make any sense to do. Shipping compilers with a game executable and stuff. At least, that's the only way I could envision such a thing being remotely possible.

> the Inject annotation specifies a method to inject into as well as where using the At annotation (HEAD (top of the method), TAIL (bottom of the method)

I think our proposal was that if a Trait defined a method with the same name as an existing method, it would result in a naming collision parse failure. We wanted Traits to be as simple as possible, a sort of copy/paste mechanic of entire properties, methods, signals, and constants. But, with the introduction of the annotation system in GDScript 2.0, I don't think it would be *completely* out of bounds to enable a same-named trait method with an `@before` or `@after` annotation to be merged with a top-layer class's implementation. It would, however, *wildly* increase the complexity of the proposed Trait system. You'd not only have to create two parse trees and merge them together at a macro level, but also dive into the details of the targeted parse tree and identify the exact parse node that rests at the end of the corresponding method call. *Then* you'd have to copy/paste the trait's parse nodes into that location in the resulting "merged" parse tree.

So...it would theoretically be possible, but I don't think anyone would go for it, because maintaining that feature would increase the complexity of the feature a *lot*, and Godot has a tendency to favor simpler, maintainable systems that provide as much utility as possible, but without incurring too much technical debt.

> and more specific things like before or after a variable or method is called in a method and which one to place the injection at if there's multiple)

I don't really understand this. If it's a class in the first place, these things should all be under class scope. It shouldn't matter where the variable or method declarations exist within the class in the resulting merged runtime class parse tree.

> the Shadow annotation allows you to use a variable as if your mixin class was inside the specified class without needing to create it yourself.

So....basically, it lets you use a private member variable? Again, all scripted properties are public. Godot has no notion of access modifiers within its scripting API (at least, not yet).

> You're also able to create your own variables, local and not, as well as methods through mixins if you wish

The original Trait Issue was opened specifically to target this use case.

> and can also overwrite a method entirely (however obviously this would be a last ditch thing, as it could break anything trying to inject into the method you overwrote).

This would violate the intent of a Trait/Mixin concept though. I think I would favor prohibiting this behavior. If you have to override an existing method, then you would need to rely on using a derived class. Or perhaps, only allow this if someone specifically forces it with a GDScript annotation? That way, if it would have been blocked by a method conflict, you can do a `@trait_force` on the method or something to have an included trait forcibly override the function present on the original script. As you stated, it could be useful for "last ditch effort" type scenarios. But I wouldn't make it the default behavior.

Anyway, what do you think @vnen?

**Comment by CammiePone** [comment]
So... Traits are the same things as what I'm proposing? A way to inject code at runtime into any method in any class anywhere?

I was told that Traits/Mixins are different from what SpongeForge called their Mixins. I was told the two weren't the same, and that SpongeForge's choice of words was confusing.

**Comment by willnationsdev** [comment]
> So... Traits are the same things as what I'm proposing? A way to inject code at runtime into any method in any class anywhere?

No, Traits just fulfill a similar use case. But they don't allow you to arbitrarily inject your code into other people's code.

> I was told that Traits/Mixins are different from what SpongeForge called their Mixins. I was told the two weren't the same, and that SpongeForge's choice of words was confusing.

This is correct. Traits/Mixins, as they are normally understood, are a way of creating a partial class that bundle property/method information together, but which don't actually define a class. You can then reference a trait from another concrete class which merges the trait's properties/methods into the main class, as if you'd copy/pasted the code.

What your proposal describes is some alternative demonic juju that invasively modifies third-party code to suit your needs, lol. From your description of its features, SpongeForge's "Mixins" are very different from the same-named features present in other languages.

---

Now, once a user-defined means of defining GDScript annotations is implemented (which I believe is a planned feature), then it wouldn't be impossible to implement SpongeForge's system as a plugin. In fact, you can probably do it now with special comments or something.

Any GDScript file can have its source code manually set to a new string and then have the script be reloaded. This means that, theoretically, you could define a system where people mark code with special comments and then a system takes that script and its special comments, parses them, uses the data contained therein to seek out and modify the source code of the target scripts, and then reload those scripts. I don't believe there is anything that would prevent you from building such a mod-friendly system as a Godot 3.2 plugin.

**Comment by CammiePone** [comment]
> I don't believe there is anything that would prevent you from building such a mod-friendly system as a Godot 3.2 plugin.
Ah, okay. I've been getting mixed things on whether or not it was doable. Friend who said traits and SpongeForge "mixins" were different said it should be possible, but I think it was Zylann (might've been someone else, my memory is awful) that said that it wasn't possible. Thanks for clarifying though. Hopefully it is possible.

**Comment by willnationsdev** [comment]
@CammiePone It isn't possible if you intend to have such a feature be part of the language, but if you want to just completely change your script code at runtime, then it is doable.

**Comment by Calinou** [comment]
Closing due to lack of support (see reactions on OP and general lack of activity here).

**Comment by thrailkillr** [comment]
> > So... Traits are the same things as what I'm proposing? A way to inject code at runtime into any method in any class anywhere?
> 
> No, Traits just fulfill a similar use case. But they don't allow you to arbitrarily inject your code into other people's code.
> 
> > I was told that Traits/Mixins are different from what SpongeForge called their Mixins. I was told the two weren't the same, and that SpongeForge's choice of words was confusing.
> 
> This is correct. Traits/Mixins, as they are normally understood, are a way of creating a partial class that bundle property/method information together, but which don't actually define a class. You can then reference a trait from another concrete class which merges the trait's properties/methods into the main class, as if you'd copy/pasted the code. But this always copy/pastes the properties/methods as an entire unit. You can't do a _part_ of a method or anything.
> 
> What your proposal describes is some alternative demonic juju that invasively modifies third-party code to suit your needs, lol. From your description of its features, SpongeForge's "Mixins" are very different from the same-named features present in other languages.
> 
> Now, once a user-defined means of defining GDScript annotations is implemented (which I believe is a planned feature), then it wouldn't be impossible to implement SpongeForge's system as a plugin. In fact, you can probably do it now with special comments or something.
> 
> Any GDScript file can have its source code manually set to a new string and then have the script be reloaded. This means that, theoretically, you could define a system where people mark code with special comments and then a system takes that script and its special comments, parses them, uses the data contained therein to seek out and modify the source code of the target scripts, and then reload those scripts. I don't believe there is anything that would prevent you from building such a mod-friendly system as a Godot 3.2 plugin. And given how specialized the behavior is / how difficult it might be to maintain such a system, I think it would be a good place for a plugin rather than the engine source code. This would, however, only apply such things to script code, i.e. having scripts that modify other GDScript files. I don't think there is any way to service your request to inject logic at the engine C++ code level.

Hello, SpongePowered Mixins are _not_ the same as Mixins in programming. A Mixin of this variety is a transformer that you can use to modify code at runtime, or mix in your own code to a method or class. Though, this seems possible anyways using a top level autoload that checks against a config file (say, a modid.mixins.json) and does some evil AST manipulation on existing scripts...
